### ISSUE TITLE
Add Optimistic updates to MUD

### ISSUE THREAD

**Issue by roninjin10** [latticexyz/mud#3133]
## TLDR

1. User uses writeOptimistic will add tx to the tevm mempool and remove it when receipt comes back
2. Under hood: Tevm will fork the current MUD stores storage and execute the tx before writing to the zustand store
3. User reads from both optimistic and cannonical state to build UI

## Api

**IMPORTANT** this api is for building a POC and is not to be considered the final API should this get merged. Feel free to comment on it however.

## Non breaking

Optimistic update support should not affect the behavior of any existing MUD apis. 

## Writing to optimistic state

By default, writes are not written to optimistic state. You must opt in.

```typescript
// this will not be reflected in optimistic state
  const tx = await worldContract.write.willRevert({ gas: 100000n });
  // this will be reflected
  const tx = await worldContract.writeOptimistic.willRevert({ gas: 100000n });
```

See
## Reading from optimistic state

By default, all reads read from cannonical state. You must opt in to reading optimistically.

```typescript
useStore.subscribe((state, prevState) => {
  const record = state.getRecordOptimistic(tables.MessageTable, {});
  if (record && record !== prevState.records[record.id]) {
    document.getElementById("chat-output")!.innerHTML += `${new Date().toLocaleString()}: ${record?.value.value}\n`;
  }
});
```

## Enabling optimistic updates

For a POC just gonna have it be always enabled. In future we could consider alternatives to optimize tree shaking or lazy loading Tevm

## Implementation of writeOptimistic

There is an existing abstraction in tevm repo that extends viem to write optimistically to tevm. We can copy paste that in.

What it is doing is doing a normal writeContract but then also submitting the tx to the tevm mempool. This makes tevm consider it a pending tx. When the receipt comes back we remove the tx from the mempool

## Implementation of getRecordOptimistic

Optimistic state is stored alongside cannonical state in the zustand store. e.g. `records` and `recordsOptimistic`.

## How tevm maintains optimistic state

Tevm treats optimistic state as derived state and calculates it on demand. If a json-rpc request has block tag 'pending' then it will mine a new block with all pending tx in it's tx pool before executing the request.

## How to optimize this process for plugging into MUD

Tevm has no responsiblity other than to derive 'pending' state on top of MUD's cannonical state via executing any tx in the optimistic tx pool

#### 1. Make a new StateManager for Tevm that uses MUD to get storage slot data

This will simplify things a lot because instead of Tevm needing to worry about syncing and cache invalidation it
can use local instance of MUD as source of truth for cannonical state. 

Since MUD storage is available locally it is quick to query. There is no additional overhead
to tevm needing to invalidate it's cache and refetch storage slots as new blocks come in. 

```typescript
import {StateManager} from '@tevm/state'

class MudManager implements StateManager {
   ... implement it to use MUD store

   // you might want to do this so you can also use the normal state manager for non mud things
  constructor(private readonly params, private readonly internalManager = new StateManager(params)) {}

  getStorage(address, slot) {
   return getStorageFromMud(this.params.mudStore) ?? this.internalManager(address, slot)
 }
}
```

#### 2. Cache invalidation

To optimize cache invalidation we want to make sure we keep track of which MUD storage slots are read from and written to. This allows us to invalidate cache and reexecute tx as new tx come into cannonical chain.

FOr a POC we could just throw away the cache and rebuild the pending block everytime a new block comes in or a new optimistic tx is added. It won't be a performant but it will still work.

#### 3. Writing to zustand store

Simply feed the events the EVM emits into the mud state to update the store. Make sure to do this in a way where we don't accidentally overwrite cannonical state though
